File: lyrics_app_android/app/src/main/java/com/maxlisui/lyrics_app/helper/SpotifyThread.kt
package com.maxlisui.lyrics_app.helper

import android.os.Process
import android.util.Log
import com.maxlisui.lyrics_app.SPOTIFY_THREAD_LOG_HELPER
import com.maxlisui.lyrics_app.SpotifySongToVoidAction
import java.util.concurrent.CountDownLatch

class SpotifyThread(private val spotifyHelper: SpotifyHelper, private val onNewSong: SpotifySongToVoidAction) : Runnable {
    override fun run() {
        Process.setThreadPriority(Process.THREAD_PRIORITY_BACKGROUND)
        try {
            var currentlyPlayingId = ""
            var isPaying = false
            while(true) {
                try {
                    val count = CountDownLatch(1)
                    spotifyHelper.getCurrentSong {
                        if(currentlyPlayingId != it.id || isPaying != it.isPlaying) {
                            currentlyPlayingId = it.id
                            isPaying = it.isPlaying
                            onNewSong(it)
                        }
                        count.countDown()
                    }
                    count.await()
                    Thread.sleep(2000)
                } catch (e: Exception) {
                    Log.e(SPOTIFY_THREAD_LOG_HELPER, "Error while refreshing song", e)
                }
            }
        } catch (ex: Exception) {
            Log.e(SPOTIFY_THREAD_LOG_HELPER, "Error while running thread", ex)
        }
    }
}

File: lyrics_app_ios/lyrics_app_ios/Helper/SpotifyThread.swift
import Foundation

class SpotifyThread {
    private var isPlaying = false
    private var currentlyPlayingId = ""
    private let spotifyHelper: SpotifyHelper
    private let onNewSong: (SpotifySong) -> Void
    
    
    init(spotifyHelper: SpotifyHelper, onNewSong: @escaping (SpotifySong) -> Void) {
        self.spotifyHelper = spotifyHelper
        self.onNewSong = onNewSong
    }
    
    func run() {
        DispatchQueue.global(qos: .background).async {
            while true {
                let semaphore = DispatchSemaphore(value: 1)
                self.spotifyHelper.getCurrentSong() { (song) in
                    if self.currentlyPlayingId != song.id || self.isPlaying != song.isPlaying {
                        self.currentlyPlayingId = song.id
                        self.isPlaying = song.isPlaying
                        self.onNewSong(song)
                        semaphore.signal()
                    }
                }
                semaphore.wait()
                sleep(2)
            }
        }
    }
}
